SocialNetwork/SocialNetwork.Core/Services/UserService.cs

```diff
@@ -1,36 +1,63 @@
 ﻿using SocialNetwork.Core.Interfaces;
 using SocialNetwork.Core.Models;
-
+using System.Text.RegularExpressions;
+using System.Xml.Linq;
 namespace SocialNetwork.Core.Services;
-
 public class UserService : IUserService
 {
-    private readonly IRepository _repository;
-
-    public UserService(IRepository repository)
+    // TODO repository
+    public List<User> Users {get; set; }
+    public UserService()
     {
-        _repository = repository;
+        Users = new List<User>();
     }
-
-    public IEnumerable<User> GetAllUsers(int skip = 0, int take = 20)
+    public List<User> GetUsers() 
     {
-        return _repository.GetAll<User>()
-            .Skip(skip)
-            .Take(take);
+        return Users;
     }
-
-    public async Task<User> AddUser(User user)
+    public User? GetUserById(int id)
     {
-        throw new NotImplementedException();
+        User? user = Users.FirstOrDefault(unit => unit.Id == id);
+        return user;
     }
-
-    public Task<User> UpdateUser(User user)
+    public User? GetUserByName(string name)
     {
-        throw new NotImplementedException();
+        User? user = Users.FirstOrDefault(unit => unit.Nickname == name);
+        return user;
     }
-
-    public Task DeleteUser(int id)
+    public void UpdateUser(User user)
     {
-        throw new NotImplementedException();
+        if (!UserIsValid(user))
+            throw new Exception("User credentials aren't valid"); // TODO own types of exceptions
+        var index = Users.IndexOf(Users.First(unit => unit.Id == user.Id));
+        Users[index] = user;
+    }
+    public void LogIn(User user)
+    {
+        user.IsLoggedIn = true;
+    }
+    public void LogOut(User user)
+    {
+        user.IsLoggedIn = false;
+    }
+    public void SignUp(User user)
+    {
+        if (!UserIsValid(user))
+            throw new Exception("User credentials aren't valid"); // TODO own types of exceptions
+        Users.Add(user);
+    }
+    // Validation logic **************************************************************************************************************************
+    private bool UserIsValid(User user)
+    {
+        // Cheks if a string has at least one latin character, at least one digit and only one '_' character. Other characters should be excluded
+        var nick_name_validation = DataIsValid(user.Nickname, @"^(?=.*[a-zA-Z])(?=.*\d)[a-zA-Z\d]*_[a-zA-Z\d]*$");
+        // Cheks if a string has at least one lower-case latin character, at least one upper-case latin character and at least one digit. The string must be at least 8 characters long
+        var password_validation = DataIsValid(user.Password, @"^(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}$");
+        return nick_name_validation && password_validation;
+    }
+    private bool DataIsValid(string data, string pattern)
```

Comment:
No need to create a separate method for two lines of code. It could be eve simplified to `private bool IsDataValid(string data, string pattern) => Regex(pattern).IsMatch(data)`